File: hash/groupAnagram.js
/*
Given an array of strings strs, group the

together. You can return the answer in any order.

 

Example 1:

Input: strs = ["eat","tea","tan","ate","nat","bat"]

Output: [["bat"],["nat","tan"],["ate","eat","tea"]]

Explanation:

    There is no string in strs that can be rearranged to form "bat".
    The strings "nat" and "tan" are anagrams as they can be rearranged to form each other.
    The strings "ate", "eat", and "tea" are anagrams as they can be rearranged to form each other.

Example 2:

Input: strs = [""]

Output: [[""]]

Example 3:

Input: strs = ["a"]

Output: [["a"]]

 */


function groupAnagram(strs){
    
	    const anagramsMap = new Map();
	    
	    strs.forEach(s => {
	        const count = new Array(26).fill(0);
	        for (let c of s) {
	            count[c.charCodeAt(0) - 'a'.charCodeAt(0)]++;
	        }
	        const key = count.join(',');
	        if (!anagramsMap.has(key)) {
	            anagramsMap.set(key, []);
	        }
	        anagramsMap.get(key).push(s);
	    });
	    
	    return Array.from(anagramsMap.values());

}

let arr = ["eat","tea","tan","ate","nat","bat"];

console.log(groupAnagram(arr));

File: hash/groupAnagram.cpp
#include<iostream>
#include<unordered_map>
#include<string>
#include<vector>
#include<algorithm>
#include<sstream>
#include<iterator>

std::string getKey(std::string& word){
    std::vector<int> key(26,0);

    for(int c = 0;c<word.size();c++){
        int ascii = word[c] - 97;
        key[ascii] = key[ascii] + 1;
    }

    std::stringstream result;
    std::copy(key.begin(),key.end(),std::ostream_iterator<int>(result,""));
    
    return result.str();
}

void groupAnagram(std::vector<std::string>& words){
    std::unordered_map<std::string,std::vector<std::string>> anagrams;

    for(int i = 0;i<words.size();i++){
        std::string key = getKey(words[i]);
        anagrams[key].push_back(words[i]);
    }
   
    for (const auto& [key, value] : anagrams) {
        std::cout << "{";
        for(const auto& word : value){
            std::cout << " '" << word << "' ";
        }
        std::cout << "}" << std::endl;
    }

}
int main(){
    std::vector<std::string> words = {"eat","tea","tan","ate","nat","bat"};
    groupAnagram(words);
    return 0;
}